spring-cloud-zookeeper-discovery/src/main/java/org/springframework/cloud/zookeeper/discovery/ZookeeperServerList.java

```diff
@@ -86,7 +88,14 @@ private String getServiceIdFromDepsOrClientName(IClientConfig clientConfig, Zook
 			}
 			List<ZookeeperServer> servers = new ArrayList<>();
 			for (ServiceInstance<ZookeeperInstance> instance : instances) {
-				servers.add(new ZookeeperServer(instance));
+				String instanceStatus = null;
+				if (instance.getPayload() != null && instance.getPayload().getMetadata() != null) {
+					instanceStatus = instance.getPayload().getMetadata().get(INSTANCE_STATUS_KEY);
+				}
+				if (!StringUtils.hasText(instanceStatus) // backwards compatibility
```

Comment:
I think this guards against lack of payload properly. But it would be good to do some test for this I guess.